### DISCUSSION TITLE
Auto-cast dictionaries to declared types of specific attributes in a model

### DISCUSSION THREAD

**Discussion by microchipster** [pydantic/pydantic#2877] [Question] [answered]
Hi,

I'm not sure what the right terminology for what I want to do is, but I think pydantic might be able to help.

I want to read a dictionary from a json/yaml file and cast the dictionary to a `BaseModel`, which will validate the types. But I also want to cast dictionaries matching the attribute name to any declared nested types to have those validated too. An example with one layer of nesting:

```
from pydantic import BaseModel

class TopClass(BaseModel):
    name: str
    items: Sequence[int]
    sub_object: NestedObject

    class Config:
        frozen = True
        validate_all = True
        validate_assignment = True
        extra = "forbid"

class NestedObject(BaseModel):
    name: str
    other_attribute: int

    class Config:
        frozen = True
        validate_all = True
        validate_assignment = True
        extra = "forbid"

# read from json
data = {
    "name": "some name",
    "items": [0, 1, 2],
    "sub_object": {"name": "sub name", "other_attribute": 3}
}

# want this to auto cast data["sub_object"] to NestedObject
top_class = TopClass(**data)
```

So could do this, which isn't terrible:
```
data["sub_object"] = SubObject(**data["sub_object"])
top_class = TopClass(**data)
```
But I am wondering if there is a way to automate this because it can get verbose if there are many layers of nested types.

**Comment by microchipster** [comment]
Figured out that I can get the caller syntax I wanted by adding this to TopClass, which isn't as bad, but still curious if I'm missing something easier:
```
    @validator("sub_object")
    @classmethod
    def valid_sub_object(cls, sub_object: Union[Mapping, NestedObject]) -> Sequence[NestedObject]:
        """ Validate and cast sub_object attribute """
        if isinstance(sub_object, dict):
            return NestedObject(**sub_object)
        return sub_object
```

**Comment by nymous** [comment, marked as answer]
I am not sure I understand your issue, because your code works, and it will parse your dict recursively in the models you want.

```py
from typing import Sequence

from pydantic import BaseModel

class NestedObject(BaseModel):
    name: str
    other_attribute: int

    class Config:
        frozen = True
        validate_all = True
        validate_assignment = True
        extra = "forbid"

class TopClass(BaseModel):
    name: str
    items: Sequence[int]
    sub_object: NestedObject

    class Config:
        frozen = True
        validate_all = True
        validate_assignment = True
        extra = "forbid"

# read from json
data = {
    "name": "some name",
    "items": [0, 1, 2],
    "sub_object": {"name": "sub name", "other_attribute": 3}
}

# want this to auto cast data["sub_object"] to NestedObject
top_class = TopClass(**data)

print(top_class)
print(top_class.sub_object)
```

I moved the `NestedObject` class before the `TopClass` because of references issues but that's the only change.

**Reply by microchipster** [reply]
Oh wow, :facepalm: , that does work! It even works if you have a list of the nested objects. I had convinced myself that I needed to do something more with pydantic.Field or something like that before even trying it.

Well thanks so much for clarifying that! Great project.
